allow to suppress logging of completed requests

bbesser opened the pull request:
Added a flag to turn off logging of completed requests.

gravis commented:
Same comment as https://github.com/meatballhat/negroni-logrus/pull/35#issuecomment-498254921 :)

bbesser commented:
Hi, thanks for your reply. The intention is similar as for logStarting=false, to replace the generic "completed"-message with a more expressive one for the particular request. In the end, this drastically reduces the log volume due to 'duplicated' messages.

gravis commented:
Ok, so the user would have to log the "completed" event. this hook has to be the very last, and you can't easily guarantee that. At least, I think we should have a section in the README to explain that. What do you think?

bbesser commented:
With "last hook" you mean "last in the chain of middlewares", I guess. Then, yes, it is up to the user to make sure that logging takes place after all middlewares have finished.

A note in the README would be helpful, I think.

bbesser commented:
Added section to the readme.

gravis commented on README.md:
```diff
@@ -14,3 +14,13 @@ custom log level on the logger with `NewCustomMiddleware`.
+You are free to disable these log messages using the options `logStarting` and `logCompleted` of the `Middleware` struct.
+The intention is to allow to replace these generic messages with custom messages more fitted to the usecase.
+
+Note that if logging completed requests is disabled and negroni-logrus is part of an entire chain of middlewares, logging the exact moment of the completed request is not possible.
```
I'm starting to think this is over-engineered. We'll end-up in messy code in the final apps with a risk of misplacing the last log call.
Since we already have an `middleware.After` [function](https://github.com/meatballhat/negroni-logrus/blob/master/middleware.go#L141), it should be used to achieve what you want.
It would be a breaking change, but what about move `[...].Info("completed handling request")
` to [`DefaultAfter`](https://github.com/meatballhat/negroni-logrus/blob/master/middleware.go#L162), so users can provide a complete function to achieve exactly what they want. It would be a cleaner code I think.
What do you think?
(sorry for the change of direction, I haven't touched this project since a while, and I forgot about this func).

bbesser replied:
The change is breaking for people who provide a custom `middleware.After` function and rely on the existing `[...].Info("completed handling request")`. For those people, completed-messages will be gone.

My feeling is that people who replace `After` also log custom messages, i.e., the change is 'not so breaking'. Also, fixing the change is easy by simply adding an `Info(...)` to the custom `After`.

Imo, that change is ok.

gravis replied:
I'm fine with a breaking change like this, it's just a matter of versioning after all. As you said, it will just remove the "completed handling request" text. My question was more, are you ok changing the behavior of this PR accordingly? Thanks

bbesser replied:
Will do, however not until next week. Will report back.

bbesser commented:
Hi, moved completion logging to DefaultAfter and changed the README.md accordingly. It's already a breaking change, so now I wonder if--for consistency--start logging should be moved to DefaultBefore.

gravis commented on README.md:
```diff
@@ -14,3 +14,9 @@ custom log level on the logger with `NewCustomMiddleware`.
+
+## Logging
+
+By default, starting requests and completed requests will be logged.
```
I'd like to change this, as it seems we're talking about two requests here. 

```suggestion
By default, requests will generate a log entry when starting and completing.
```

gravis commented on middleware.go:
```diff
@@ -160,10 +160,11 @@ func DefaultBefore(entry *logrus.Entry, req *http.Request, remoteAddr string) *l
 
 // DefaultAfter is the default func assigned to *Middleware.After
 func DefaultAfter(entry *logrus.Entry, res negroni.ResponseWriter, latency time.Duration, name string) *logrus.Entry {
-	return entry.WithFields(logrus.Fields{
```
I don't think calling `WithFields` or `Info()` will actually change the entry. Is there a reason to keep the `return entry`?

bbesser replied:
I was hesitating to remove the return and change `After`'s signature, since this is a compile incompatibility. Otherwise, there is no reason to keep it. Even more, the compiler would really point users to the change :-D

bbesser replied:
So, let me know if changing the signature is ok for you. If so, i can make the change.

gravis replied:
Ok, let's do that if you don't mind. Thanks

gravis commented on middleware_test.go:
```diff
@@ -178,12 +178,7 @@ func TestMiddleware_ServeHTTP_AfterOverride(t *testing.T) {
-			`"remote":"10.10.10.10","request":"http://example.com/stuff?rly=ya",`+
-			`"request_id":"22035D08-98EF-413C-BBA0-C4E66A11B28D","time":"%s"}`, nowToday),
-		lines[1])
+	assert.Len(t, lines, 1)
```
This test is confusing now, we should have an entry with `{"hambone": 57}`.
Which makes me think again that we shouldn't return anything.

bbesser replied:
True. The test seemed alright for me, since overriding `After` leaves only one log line (for the request start). However, line 175 really is confusing as hell.

gravis replied:
Once we change the signature of `*Middleware.After`, this will make sense again.

gravis commented:
@bbesser I got some negative feedback about my choice here. Some of our users still don't use GOMODULES everywhere, and breaking the API will be an issue. I reverted the changes, and to avoid asking you more work, I created a MR with your first commit to merge it directly: https://github.com/meatballhat/negroni-logrus/pull/38
I hope it's fine with you. I'm going to close this PR, and will merge the other instead once tests are passing.
Thanks again for your contribution, it's really appreciated.
